File: java/src/main/java/com/aurelion/connector/Service.java
package com.aurelion.connector;

import com.aurelion.connector.core.OperationExecutor;
import com.aurelion.connector.core.Storage;

import java.util.*;

public final class Service implements OperationExecutor {

    private final Backend backend;
    private final Map<String, Op> registry;

    @FunctionalInterface
    private interface Op {
        OperationResult apply(Map<String, Object> payload, boolean resultStorageRequested,
                               String correlationId);
    }

    public Service(String dbPath) {
        this(dbPath, true);
    }

    public Service(String dbPath, boolean autoSeedMock) {
        this.backend = new Backend(dbPath);
        this.backend.initDb(autoSeedMock);
        Map<String, Op> r = new LinkedHashMap<>();
        r.put("create_account", (p, rsr, cid) -> createAccount(p));
        r.put("delete_account", (p, rsr, cid) -> deleteAccount(p));
        r.put("list_accounts",  (p, rsr, cid) -> listRecords("accounts",   backend.listAccounts(), rsr, cid));
        r.put("list_roles",     (p, rsr, cid) -> listRecords("roles",      backend.listRoles(), rsr, cid));
        r.put("list_privileges",(p, rsr, cid) -> listRecords("privileges", backend.listPrivileges(), rsr, cid));
        r.put("list_persons",   (p, rsr, cid) -> listRecords("persons",    backend.listPersons(), rsr, cid));
        r.put("list_employments",(p, rsr, cid) -> listRecords("employments", backend.listEmployments(), rsr, cid));
        this.registry = Collections.unmodifiableMap(r);
    }

    @Override
    public OperationResult execute(String operation, Map<String, Object> payload,
                                   boolean resultStorageRequested, String correlationId) {
        Op op = registry.get(operation);
        if (op == null) {
            throw new IllegalArgumentException("unsupported operation: '" + operation + "'");
        }
        return op.apply(payload, resultStorageRequested, correlationId);
    }

    private OperationResult createAccount(Map<String, Object> payload) {
        String username = asString(payload, "username");
        String email = asString(payload, "email");
        if (username == null || username.isBlank()) throw new IllegalArgumentException("username is required");
        if (email == null || email.isBlank()) throw new IllegalArgumentException("email is required");

        backend.insertAccount(username, email);
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("username", username);
        out.put("email", email);
        return new OperationResult(out, null);
    }

    private OperationResult deleteAccount(Map<String, Object> payload) {
        String username = asString(payload, "username");
        if (username == null || username.isBlank()) throw new IllegalArgumentException("username is required");

        backend.deleteAccount(username);
        return new OperationResult(Map.of("username", username), null);
    }

    private OperationResult listRecords(String datasetType,
                                         List<Map<String, Object>> records,
                                         boolean resultStorageRequested,
                                         String correlationId) {
        if (resultStorageRequested) {
            Map<String, String> ref = Storage.writeRecords(datasetType, records, correlationId);
            return new OperationResult(null, ref);
        }
        return new OperationResult(records, null);
    }

    private static String asString(Map<String, Object> map, String key) {
        Object val = map.get(key);
        return val instanceof String s ? s : null;
    }
}

File: python/service.py
"""Connector instance operation implementations."""

from typing import Any

from backend import (
    delete_account,
    init_db,
    insert_account,
    list_accounts,
    list_employments,
    list_persons,
    list_privileges,
    list_roles,
)
from core.storage import write_records

OperationResult = tuple[
    dict[str, Any] | list[dict[str, Any]] | None,
    dict[str, str] | None,
]


class Service:
    def __init__(
        self,
        db_path: str | None = None,
        *,
        auto_seed_mock: bool = True,
    ) -> None:
        self._db_path = db_path
        init_db(db_path, auto_seed=auto_seed_mock)
        self._registry: dict[str, Any] = {
            'create_account': self._create_account,
            'delete_account': self._delete_account,
            'list_accounts': self._list_accounts,
            'list_roles': self._list_roles,
            'list_privileges': self._list_privileges,
            'list_persons': self._list_persons,
            'list_employments': self._list_employments,
        }

    def execute(
        self,
        operation: str,
        payload: dict[str, Any],
        result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        handler = self._registry.get(operation)
        if handler is None:
            raise ValueError(f'unsupported operation: {operation!r}')
        return handler(payload, result_storage_requested, correlation_id=correlation_id)

    def _create_account(
        self,
        payload: dict[str, Any],
        _result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        username = payload.get('username')
        email = payload.get('email')
        if not isinstance(username, str) or not username:
            raise ValueError('username is required')
        if not isinstance(email, str) or not email:
            raise ValueError('email is required')

        insert_account(username, email, path=self._db_path)
        return {'username': username, 'email': email}, None

    def _delete_account(
        self,
        payload: dict[str, Any],
        _result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        username = payload.get('username')
        if not isinstance(username, str) or not username:
            raise ValueError('username is required')

        delete_account(username, path=self._db_path)
        return {'username': username}, None

    def _list_records(
        self,
        dataset_type: str,
        records: list[dict[str, Any]],
        result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        if result_storage_requested:
            return None, write_records(dataset_type, records, correlation_id=correlation_id)
        return records, None

    def _list_accounts(
        self,
        _payload: dict[str, Any],
        result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        return self._list_records(
            'accounts',
            list_accounts(path=self._db_path),
            result_storage_requested,
            correlation_id=correlation_id,
        )

    def _list_roles(
        self,
        _payload: dict[str, Any],
        result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        return self._list_records(
            'roles',
            list_roles(path=self._db_path),
            result_storage_requested,
            correlation_id=correlation_id,
        )

    def _list_privileges(
        self,
        _payload: dict[str, Any],
        result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        return self._list_records(
            'privileges',
            list_privileges(path=self._db_path),
            result_storage_requested,
            correlation_id=correlation_id,
        )

    def _list_persons(
        self,
        _payload: dict[str, Any],
        result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        return self._list_records(
            'persons',
            list_persons(path=self._db_path),
            result_storage_requested,
            correlation_id=correlation_id,
        )

    def _list_employments(
        self,
        _payload: dict[str, Any],
        result_storage_requested: bool,
        *,
        correlation_id: str | None = None,
    ) -> OperationResult:
        return self._list_records(
            'employments',
            list_employments(path=self._db_path),
            result_storage_requested,
            correlation_id=correlation_id,
        )
